packages/boot-contract-derive/src/lib.rs

```diff
@@ -0,0 +1,59 @@
+#![recursion_limit = "128"]
+
+extern crate proc_macro;
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, AttributeArgs, Fields, Item, Meta, NestedMeta, Path};
+
+#[proc_macro_attribute]
+pub fn boot_contract(attrs: TokenStream, input: TokenStream) -> TokenStream {
+    let mut item = parse_macro_input!(input as syn::Item);
+    let attributes = parse_macro_input!(attrs as AttributeArgs);
+
+    let types_in_order: Vec<Path> = attributes
+        .into_iter()
+        .map(|attr| {
+            let NestedMeta::Meta(Meta::Path(type_path)) = attr else {
+            panic!("Expected a contract endpoint type.");
+        };
+            type_path
+        })
+        .collect();
+
+    let exec = types_in_order[0].clone();
+    let init = types_in_order[1].clone();
+    let query = types_in_order[2].clone();
+    let migrate = types_in_order[3].clone();
```

Comment:
Is this ordering "enforced" by the type usages on lines #51-57?

Reply:
Not yet, we could add a check for the name of the type if we want to enforce that.

Reply:
But a user would quickly see that they gave the types in the wrong order as a `.instantiate()` call would ask for their `QueryMsg` or whatever